pytorch_lightning/utilities/cli.py

```diff
@@ -0,0 +1,274 @@
+        <https://omni-us.github.io/jsonargparse/#jsonargparse.core.ArgumentParser.__init__>`_.
+        """
+        super().__init__(*args, parse_as_dict=parse_as_dict, **kwargs)
+        self.add_argument(
+            '--config',
+            action=ActionConfigFile,
+            help='Path to a configuration file in json or yaml format.'
+        )
+
+    def add_trainer_args(
+        self,
+        trainer_class: Type[Trainer] = Trainer,
+        nested_key: str = 'trainer'
+    ):
+        """
+        Adds arguments from a trainer class to a nested key of the parser
+
+        Args:
+            trainer_class: Optional extension of the Trainer class.
+            nested_key: Name of the nested namespace to store arguments.
+        """
+        assert issubclass(trainer_class, Trainer)
+        return self.add_class_arguments(trainer_class, nested_key)
+
+    def add_module_args(
+        self,
+        module_class: Type[LightningModule],
+        nested_key: str = 'module',
+        subclass_mode: bool = False
+    ):
+        """
+        Adds arguments from a module class to a nested key of the parser
+
+        Args:
+            module_class: A LightningModule class.
+            nested_key: Name of the nested namespace to store arguments.
+            subclass_mode: Whether allow any subclass of the given class.
+        """
+        assert issubclass(module_class, LightningModule)
+        if subclass_mode:
+            return self.add_subclass_arguments(module_class, nested_key)
+        return self.add_class_arguments(module_class, nested_key)
+
+    def add_datamodule_args(
+        self,
+        datamodule_class: Type[LightningDataModule],
+        nested_key: str = 'data',
+        subclass_mode: bool = False
+    ):
+        """
+        Adds arguments from a datamodule class to a nested key of the parser
+
+        Args:
+            datamodule_class: A LightningDataModule class.
+            nested_key: Name of the nested namespace to store arguments.
+            subclass_mode: Whether allow any subclass of the given class.
+        """
+        assert issubclass(datamodule_class, LightningDataModule)
+        if subclass_mode:
+            return self.add_subclass_arguments(datamodule_class, nested_key)
+        return self.add_class_arguments(datamodule_class, nested_key)
+
+
+class SaveConfigCallback(Callback):
+    """Saves a LightningCLI config to the log_dir when training starts"""
+
+    def __init__(self, parser, config):
+        self.parser = parser
+        self.config = config
+
+    def on_train_start(self, trainer, pl_module):
+        config_path = os.path.join(trainer.logger.log_dir, 'config.yaml')
+        self.parser.save(self.config, config_path, skip_none=False)
+
+
+class LightningCLI:
+    def __init__(
+        self,
+        model_class: Type[LightningModule],
+        datamodule_class: Type[LightningDataModule] = None,
+        save_config_callback: Type[Callback] = SaveConfigCallback,
```

Comment:
why it is needed here as type?

Reply:
Type is required because the argument must be the class. Not an instance of the class.

Reply:
I changed the type to be `Type[SaveConfigCallback]` since it is not any callback that can be given, it must be a callback that is instantiated with `parser` and `config`.